src/google/protobuf/json/internal/parser.cc

```diff
@@ -625,6 +625,180 @@ absl::Status ParseArray(JsonLexer& lex, Field<Traits> field, Msg<Traits>& msg) {
+    LocationWith<MaybeOwnedString>& key) {
+  // Parse the enum value from string, advancing the lexer.
+  absl::StatusOr<absl::optional<int32_t>> enum_value;
+  (void)Traits::WithFieldType(
+      map_field, [&lex, &enum_value](const Desc<Traits>& map_entry_desc) {
+        enum_value = ParseEnum<Traits>(lex, Traits::ValueField(map_entry_desc));
+        return absl::OkStatus();
+      });
+  RETURN_IF_ERROR(enum_value.status());
+
+  // If enum value is unknown, we'll stop here and avoid allocating a new map
+  // entry message.
+  if (!enum_value->has_value()) {
+    return absl::OkStatus();
+  }
+
+  // If the enum value is known, output the map entry with already parsed
+  // value.
+  return Traits::NewMsg(
+      map_field, parent_msg,
+      [&](const Desc<Traits>& type, Msg<Traits>& entry) -> absl::Status {
+        RETURN_IF_ERROR(ParseMapKey<Traits>(lex.options(), type, entry, key));
+
+        // This is different from the default implementation in 'ParseMapEntry'
+        // since we already parsed this value (we must not attempt to advance
+        // the lexer to parse the value here again).
+        Traits::SetEnum(Traits::ValueField(type), entry,
+                        enum_value->value_or(0));
+
+        return absl::OkStatus();
+      });
+}
+
+// ParseMapEntry implementation has a special case if:
+// * 'map_field' is a map of enums, and
+// * value that follows in 'lex' is a string (enum name), and
+// * and ignore_unknown_fields is set.
+// This function does not advance the lexer.
+template <typename Traits>
+absl::StatusOr<bool> MapEntryNeedsEnumStringValueSpecialCase(
+    JsonLexer& lex, Field<Traits> map_field) {
+  // Check if the map_field is a map of enums.
+  bool is_map_of_enums = false;
+  (void)Traits::WithFieldType(
+      map_field, [&is_map_of_enums](const Desc<Traits>& desc) {
+        is_map_of_enums = Traits::FieldType(Traits::ValueField(desc)) ==
+                          FieldDescriptor::TYPE_ENUM;
+        return absl::OkStatus();
+      });
+  if (!is_map_of_enums) {
+    return false;
+  }
+
+  // Check that the value is a string (instead of a number).
+  absl::StatusOr<JsonLexer::Kind> kind = lex.PeekKind();
+  RETURN_IF_ERROR(kind.status());
+  if (*kind != JsonLexer::kStr) {
+    return false;
+  }
+
+  // We need the special case only if we are ignoring unknown fields.
+  return lex.options().ignore_unknown_fields;
+}
+
+// Parses one map entry for 'map_field' in 'parent_msg' with already consumed
+// 'key'.
+template <typename Traits>
+absl::Status ParseMapEntry(JsonLexer& lex, Field<Traits> map_field,
+                           Msg<Traits>& parent_msg,
+                           LocationWith<MaybeOwnedString>& key) {
+  auto needs_special_case =
+      MapEntryNeedsEnumStringValueSpecialCase<Traits>(lex, map_field);
+  RETURN_IF_ERROR(needs_special_case.status());
+  if (*needs_special_case) {
+    return ParseMapEntryWithEnumStringValueWhenIgnoringUnknownFields<Traits>(
```

Comment:
Can we structure this to be less drastically special cased? I'd rather it 'infect' the happy case if that avoids such a large diversion to handle the edge case.

Is it possible to structure this to always parse map key, parse map value, decide whether to err or ok if the value can't be parsed due to it being unknown enum, then construct the message?